### DISCUSSION TITLE
FileUpload to s3 is work, but remove record's images no delete Buckets file

### DISCUSSION THREAD

**Discussion by jsttech** [filamentphp/filament#9558] [Help] [answered]
I using v3 demo panel

```
Forms\Components\Section::make('Images')
->schema([
    SpatieMediaLibraryFileUpload::make('media')
        ->disk('s3')
        ->collection('product-images')
        ->multiple()
        ->maxFiles(5)
        ->disableLabel(),
])
->collapsible(),
```

my Bucket policy

```
{
    "Version": "2012-10-17",
    "Id": "Policy1699519102039",
    "Statement": [
        {
            "Sid": "Stmt1699519100562",
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:*",
            "Resource": "arn:aws:s3:::app/*"
        }
    ]
}
```

How can I fix this problem, thx

**Comment by leandrocfe** [comment]
*It is the responsibility of the developer to delete these files from the disk if they are removed, as Filament is unaware if they are depended on elsewhere. One way to do this automatically is observing a [model event](https://laravel.com/docs/eloquent#events).*
https://w3codegenerator.com/code-snippets/laravel/delete-file-from-amazon-s3-bucket-using-laravel

**Reply by jsttech** [reply]
local storage delete file is work

**Reply by kokoshneta** [reply]
It’s easy to simply say that it’s the developer’s responsibility. The difficulty is finding a useful way to accomplish it. Since the FileUpload component does not – as far as I can tell – provide any way at all to modify the removal process, you’re left with hacking it together.

Using a model event to delete files from S3 storage is quite wasteful, since model events are fired on _every update_ to the model. Needing to check whether one specific attribute is dirty and has been emptied _every time any attribute is ever modified_ is highly inefficient, not to mention unintuitive.

The correct place to define the behaviour for removing an uploaded file is in the FileUpload component. The BaseFileUpload class even has a `deleteUploadedFileUsing()` method, which it would make perfect sense to use… except none of the deletion methods that are available in that class seem to be used anywhere in the component.

Even using `afterStateUpdated` to conditionally remove the file if `$state` is empty doesn’t seem to do anything.

Why can the remove action not allow the developer to utilise the existing options to control file deletion?

**Reply by jsttech** [reply, marked as answer]
@kokoshneta my solution

### DefaultPathGenerator.php

```
    protected function getBasePath(Media $media): string
    {
        $prefix = Auth::user()->id. '/' . $media->model_id;

        if ($prefix !== '') {
            return $prefix . '/' . $media->getKey();
        }
        return $media->getKey();
    }
```

### InteractsWithMedia.php
```
    public static function bootInteractsWithMedia()
    {
        static::deleting(function (HasMedia $model) {
            if ($model->shouldDeletePreservingMedia()) {
                return;
            }
            if (in_array(SoftDeletes::class, class_uses_recursive($model))) {
                if (! $model->forceDeleting) {
                    return;
                }
            }
            $model->media()->cursor()
            ->each(function (Media $media): void {
                if($media->disk == 's3'){
                    $filepath = Auth::user()->id. '/' . $media->model_id. '/' . $media->id . '/' . $media->getAttributeValue('file_name');
                    if(Storage::disk('s3')->delete($filepath)){
                        $media->delete();
                    }
                }
                else { $media->delete(); }
            });
        });
    }
```

### SpatieMediaLibraryFileUpload.php

```
public function deleteAbandonedFiles(): void
    {
        /** @var Model&HasMedia $record */
        $record = $this->getRecord();
        $record
        ->getMedia($this->getCollection())
        ->whereNotIn('uuid', array_keys($this->getState() ?? []))
        ->each(function (Media $media): void {
            if($this->diskName == 's3'){
                $filepath = Auth::user()->id. '/' . $media->model_id. '/' . $media->id . '/' . $media->getAttributeValue('file_name');
                if(Storage::disk('s3')->delete($filepath)){
                    $media->delete();
                }
            }
            else { $media->delete(); }
        });
    }
```

**Comment by denizaygundev** [comment]
My solution (not specific to spatie's media library)

This will work with one or multiple files (if using `->multiple`)
```
    protected static function booted(): void
    {
        static::deleted(function (File $file) {
            Storage::disk('s3')->delete($file->url);
        });
    }
```
